openpype/plugins/publish/collect_audio.py

```diff
@@ -0,0 +1,83 @@
+import pyblish.api
+from pprint import pformat
+
+from openpype.client import (
+    get_last_version_by_subset_name,
+    get_representations,
+)
+from openpype.pipeline import (
+    legacy_io,
+    get_representation_path,
+)
+
+
+class CollectAudio(pyblish.api.InstancePlugin):
+    """ Collecting available audio subset to instance
+
+    """
+    label = "Collect Audio"
+    order = pyblish.api.CollectorOrder + 0.1
+    families = ["review"]
+    hosts = [
+        "nuke",
+        "maya",
+        "shell",
+        "hiero",
+        "premiere",
+        "harmony",
+        "traypublisher",
+        "standalonepublisher",
+        "fusion",
+        "tvpaint",
+        "resolve",
+        "webpublisher",
+        "aftereffects",
+        "flame",
+        "unreal"
+    ]
+
+    audio_subset_name = "audioMain"
+
+    def process(self, instance):
+        if instance.data.get("audio"):
+            self.log.info(
+                "Skipping Audio collecion. It is already collected"
+            )
+            return
+
+        # Add audio to instance if exists.
+        self.log.info('Collecting Audio Data ...')
+
+        project_name = legacy_io.active_project()
+        asset_name = instance.data["asset"]
+
+        # Find latest versions document
+        last_version_doc = get_last_version_by_subset_name(
+            project_name,
+            self.audio_subset_name,
+            asset_name=asset_name,
+            fields=["_id"]
+        )
+
+        repre_doc = None
+        if last_version_doc:
+            # Try to find it's representation (Expected there is only one)
+            repre_docs = list(get_representations(
+                project_name, version_ids=[last_version_doc["_id"]]
+            ))
+            if not repre_docs:
+                self.log.warning(
+                    "Version document does not contain any representations"
+                )
+            else:
+                repre_doc = repre_docs[0]
```

Comment:
In most publishes many instances will share the same asset. These database queries can be slow for many instances too. To optimize it would be nice to "cache" the queries so we'd do much fewer database queries.

```python
cache = context.data.get("__cache_asset_audio", {})
if asset in cache:
    return cache[asset]

# Else query as usual and store in cache
```

References where this is also done:
- https://github.com/pypeclub/OpenPype/blob/4a4bb22f60353bb9cfcdb64c516b2b145cb7c966/openpype/hosts/maya/plugins/publish/collect_inputs.py#L116-L126

The alternative option would be to turn this in a ContextPlugin so it runs once and collects for all instances manually

Reply:
good point, thanks